src/components/lib/sl/sl_sched.c

```diff
@@ -166,12 +167,58 @@ sl_thd_block_no_cs(struct sl_thd *t, sl_thd_state_t block_type, cycles_t timeout
 		goto update;
 	}
 
-	assert(t->state == SL_THD_RUNNABLE);
+	assert(t->state == SL_THD_RUNNABLE || t->state == SL_THD_WOKEN);
 	sl_mod_block(sl_mod_thd_policy_get(t));
 
 update:
 	t->state = block_type;
 	if (block_type == SL_THD_BLOCKED_TIMEOUT) sl_timeout_block(t, timeout);
+	t->sched_blocked = 1;
+
+	return 0;
+}
+
+/*
+ * Wake "t" up if it was previously blocked on cos_rcv and got
+ * to run before the scheduler (tcap-activated)!
+ */
+static inline int
+sl_thd_sched_unblock_no_cs(struct sl_thd *t)
+{
+	if (unlikely(!t->sched_blocked)) return 0;
```

Comment:
I'll make my comments be more clear and broader to cover the states in `sched_blocked`. 
`sched_blocked` is a state variable for kernel-scheduler notifications. At the user-level, any states are possible when `sched_blocked` is "still" set because a thread may get activated by TCap mechanism and it could block/yield before the scheduler runs.